link_resolver_betterttv.go

```diff
@@ -0,0 +1,156 @@
+		type BTTVEmoteAPIResponse struct {
+			ID             string    `json:"id"`
+			Code           string    `json:"code"`
+			ImageType      string    `json:"imageType"`
+			CreatedAt      time.Time `json:"createdAt"`
+			UpdatedAt      time.Time `json:"updatedAt"`
+			Global         bool      `json:"global"`
+			Live           bool      `json:"live"`
+			Sharing        bool      `json:"sharing"`
+			ApprovalStatus string    `json:"approvalStatus"`
+			User           struct {
+				ID          string `json:"id"`
+				Name        string `json:"name"`
+				DisplayName string `json:"displayName"`
+				ProviderID  string `json:"providerId"`
+			} `json:"user"`
+		}
+
+		apiURL := fmt.Sprintf(betterttvEmoteAPIURL, emoteHash)
+		thumbnailURL := fmt.Sprintf(betterttvThumbnailFormat, emoteHash)
+
+		req, err := http.NewRequest("GET", apiURL, nil)
+		if err != nil {
+			return &LinkResolverResponse{
+				Status:  http.StatusInternalServerError,
+				Message: "betterttv http request creation error " + clean(err.Error()),
+			}, nil, noSpecialDur
+		}
+		req.Header.Set("User-Agent", "chatterino-api-cache/1.0 link-resolver")
+
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			return &LinkResolverResponse{
+				Status:  http.StatusInternalServerError,
+				Message: "betterttv http request error " + clean(err.Error()),
+			}, nil, noSpecialDur
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode < http.StatusOK || resp.StatusCode > http.StatusMultipleChoices {
+			return noBetterTTVEmoteWithThisHashFound, nil, noSpecialDur
+		}
+
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return &LinkResolverResponse{
+				Status:  http.StatusInternalServerError,
+				Message: "betterttv http body read error " + clean(err.Error()),
+			}, nil, noSpecialDur
+		}
+
+		var jsonResponse BTTVEmoteAPIResponse
+		if err := json.Unmarshal(body, &jsonResponse); err != nil {
+			return &LinkResolverResponse{
+				Status:  http.StatusInternalServerError,
+				Message: "betterttv api unmarshal error " + clean(err.Error()),
+			}, nil, noSpecialDur
+		}
+
+		data := betterttvEmoteTooltipData{
+			Code:     jsonResponse.Code,
+			Uploader: jsonResponse.User.DisplayName,
+		}
+
+		var tooltip bytes.Buffer
+		if err := tooltipTemplate.Execute(&tooltip, data); err != nil {
+			return &LinkResolverResponse{
+				Status:  http.StatusInternalServerError,
+				Message: "youtube template error " + clean(err.Error()),
+			}, nil, noSpecialDur
+		}
+
+		return &LinkResolverResponse{
+			Status:    200,
+			Tooltip:   tooltip.String(),
+			Thumbnail: thumbnailURL,
+		}, nil, noSpecialDur
+	}
+
+	cache := newLoadingCache("betterttv_emotes", load, 1*time.Hour)
+	emotePathRegex := regexp.MustCompile(`/emotes/([a-f0-9]+)`)
+
+	// Find clips that look like https://clips.twitch.tv/SlugHere
+	customURLManagers = append(customURLManagers, customURLManager{
+		check: func(url *url.URL) bool {
+			if !strings.HasSuffix("."+url.Host, ".betterttv.com") {
```

Comment:
Maybe put a small comment explaining this magic, I remember you giving a short explanation of why you used this

Reply:
The idea was to ensure we would match both www.betterttv.com and betterttv.com.
Writing this down in a comment made me realize that we would also be matching kkona.betterttv.com which is not something we wanted to do, so I replaced this with a simpler "golang set" of domains